File: backend/src/utils/ttml2ass.js
#!/usr/bin/env node
// VERSION 2.0.0
const parser = require('fast-xml-parser');

const HEADER = `[Script Info]
; This is an Advanced Sub Station Alpha v4+ script.
Title: {{TITLE}}
ScriptType: v4.00+
Collisions: Normal
PlayDepth: 0

[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
Style: Default,Arial,20,&H00FFFFFF,&H0300FFFF,&H00000000,&H02000000,0,0,0,0,100,100,0,0,1,2,1,2,10,10,10,1\n`;

const EVENT_HEADER = `\n\n[Events]\nFormat: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text\n`;

const COLORS = { YELLOW: "&H0000FFFF", CYAN: "&H00FFFF00", LIME: "&H00FFDD00", WHITE: "&H00FFFFFF", BLACK: "&H02000000" };

const STYLE_DICT = {};

const clean = (text) => {
  const LEGACY_MAPPINGS = [
    [/<span tts:color="white">/gi, '{\\c&HFFFFFF&}'],
    [/<span tts:color="yellow">/gi, '{\\c&H00FFFF&}'],
    [/<span tts:color="cyan">/gi, '{\\c&HFFFF00&}'],
    [/<span tts:color="lime">/gi, '{\\c&HFFDD00&}'],
    [/<\/span>/gi, '{\\c}'],
  ];

  const MAPPINGS = [
    [/<br \/><\/p>/gi, '</p>'],
    [/<br \/>/gi, '\\N'],
    [/&quot;/gi, '\"']
  ];

  const legacy = !text.match("urn:ebu:tt:distribution:2018-04");
  if (legacy) {
    console.log('Applying legacy settings');
    LEGACY_MAPPINGS.forEach(([entity, char]) => {
      text = text.replace(entity, char);
    });
  }

  MAPPINGS.forEach(([entity, char]) => {
    text = text.replace(entity, char);
  });

  return text;
};

const convertTimestamp = timeStr => {
  if (timeStr.length === 8) {
    return timeStr + ".00";
  } else if (timeStr.length === 12) {
    return timeStr.substring(0, 11);
  } else {
    return timeStr;
  }
};

const convertColor = colorStr => {
  if (colorStr.match(/#[a-f\d]{8}/)) {
    // TTML: Red-Green-Blue-Alpha
    // <style tts:backgroundColor="#000000ff" tts:color="#ffff00ff" xml:id="S5" />
    // ASS:  Alpha-Blue-Green-Red
    // Style: S5,Arial,20,&HFFFF00FF,&H0300FFFF,&H00000000,&HFF000000,0,0,0,0,100,100,0,0,1,2,1,2,10,10,10,1
    const match = colorStr.toUpperCase().match(/#(?<red>.{2})(?<green>.{2})(?<blue>.{2})(?<opacity>.{2})/);
    const { opacity, red, green, blue } = match.groups;
    let alpha = (255 - parseInt(opacity, 16)).toString(16);
    if (alpha.length < 2) {
      alpha = "0" + alpha;
    }

    return `&H${alpha}${blue}${green}${red}`;
  } else {
    const upColor = colorStr.toUpperCase();
    return COLORS[upColor] || upColor;
  }
};

const formatStyle = (style) => {
  // if (style['@_xml:id'].length > 5) {
  //   return "";
  // }

  const primColor = (style['@_tts:color']) ? convertColor(style['@_tts:color']) : '&H00FFFFFF';
  const backColor = (style['@_tts:backgroundColor']) ? convertColor(style['@_tts:backgroundColor']) : '&H02000000';
  STYLE_DICT[style['@_xml:id']] = primColor;

  return `Style: ${style['@_xml:id'] || style['@_id']},Arial,20,${primColor},&H0300FFFF,&H00000000,${backColor},0,0,0,0,100,100,0,0,1,2,1,2,10,10,10,1\n`;
};

const formatStyles = (styles, title) => {
  let out = HEADER.replace("{{TITLE}}", title);

  console.log('styles', JSON.stringify(styles, null, 4));
  if (Array.isArray(styles)) {
    styles.forEach(style => {
      out += formatStyle(style);
    });
  } else {
    out += formatStyle(styles);
  }

  return out;
};

const formatLine = (line) => {
  if (!line['@_begin']) {
    // if there is no timecode in the line, copy it from first nested span
    if (line['span'] && line['span'][0] && line['span'][0]['@_begin']) {
      line['@_begin'] = line['span'][0]['@_begin'];
      line['@_end'] = line['span'][0]['@_end'];
    } else {
      // error, no timecode in nested span, skip line
      return "";
    }
  }

  const begin = convertTimestamp(line['@_begin']);
  const end = convertTimestamp(line['@_end']);

  let text;
  let style = line['@_style'] || 'Default';

  if (Array.isArray(line.span)) {
    // Multiple span tags
    style = line.span[0]['@_style'] || 'Default';
    // Check if colour in all spans is the same
    const allSame = line.span.every(s => s['@_style'] === style);
    if (allSame) {
      // no need for extra colouring
      text = line.span.map(sp => sp['#text']).join('\\N');
    } else {
      // each bit with colour tags
      text = line.span.map(sp => `{\\c${STYLE_DICT[sp['@_style']]}&}${sp['#text']}{\\c}`).join('\\N');
    }
  } else if (line.span) {
    style = line.span['@_style'] || 'Default';
    text = line.span["#text"];
  } else {
    text = line["#text"];
  }
  const ass = `Dialogue: 0,${begin},${end},${style},,0,0,0,,${text}\n`;
  // console.log('line', line);
  // console.log('ass', ass);
  return ass;
};

const ttml2ass = (ttml, title) => {
  const cleanedTTML = clean(ttml);
  const json = parser.parse(cleanedTTML, { ignoreAttributes: false });
  const lines = json.tt.body.div.p;
  const styles = json.tt.head.styling.style;
  let out = formatStyles(styles, title) + EVENT_HEADER;

  lines.forEach((line) => {
    out += formatLine(line);
  });

  return out;
};

module.exports = ttml2ass;

File: backend/src/utils/ttml2ass.py
#!/usr/bin/env python3
# from: https://gist.github.com/sfan5/2f94fefbc326052b185ec995248a9ffc
import sys
from xml.dom.minidom import parseString

COLORS = { # TODO more colors
	"white"  : "FFFFFF",
	"black"  : "000000",
	"yellow" : "FFFF00",
	"green"  : "00FF00",
	"cyan"   : "00FFFF",
}

def compat_read(filename):
	f = open(filename, "rb")
	tmp = f.read(2)
	# detect UTF-16 via BOM
	if tmp == b"\xff\xfe":
		enc = "UTF-16LE"
	elif tmp == b"\xfe\xff":
		enc = "UTF-16BE"
	else:
		f.seek(0, 0)
		enc = "utf8"
	tmp = f.read().decode(enc)
	f.close()
	return tmp

def iterate_childelems(x):
	for xn in x.childNodes:
		if xn.nodeType == 3:
			continue
		yield xn

def convert_ts(s):
	vals = tuple(int(e) for e in s.split(":"))
	return "%d:%02d:%02d.%02d" % vals


if len(sys.argv) < 2 or sys.argv[1] in ("-h", "--help"):
	print("Usage: %s <input file> [output file]" % sys.argv[0], file=sys.stderr)
	print("Converts TTML subtitles into ASS format", file=sys.stderr)
	exit(1)

if len(sys.argv) > 2:
	of = open(sys.argv[2], "w")
else:
	of = sys.stdout


of.write("""[Script Info]
Title:
ScriptType: v4.00+
WrapStyle: 0

[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
""")

x = parseString(compat_read(sys.argv[1]))
assert x.childNodes[0].tagName == "tt"
tmp = list(iterate_childelems(x.childNodes[0]))
assert len(tmp) == 2
xhead, xbody = tmp[0], tmp[1]

# read style information
for xn in iterate_childelems(xhead):
	if xn.tagName != "styling":
		continue
	assert len(xn.childNodes) == 1, "Only one style supported"
	xn = xn.childNodes[0]

	fmt = "Style: Default,%s,%d,&H00%s,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,2,%d,10,10,10,1\n"
	color = COLORS[xn.getAttribute("tts:color")]
	alignment = ({"left": 1, "center": 2, "right": 3})[xn.getAttribute("tts:textAlign")]
	font = xn.getAttribute("tts:fontFamily")
	fontsz = int(xn.getAttribute("tts:fontSize"))
	of.write(fmt % (font, fontsz, color, alignment))

of.write("""
[Events]
Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text
""")

for xn in iterate_childelems(xbody):
	assert xn.tagName == "div"
	for xn in iterate_childelems(xn):
		assert xn.tagName == "p"
		ts_start = convert_ts(xn.getAttribute("begin"))
		ts_end = convert_ts(xn.getAttribute("end"))
		of.write("Dialogue: 0,%s,%s,Default,,0,0,0,," % (ts_start, ts_end))
		for xtext in xn.childNodes:
			if xtext.nodeType == 3: # raw text
				of.write(xtext.wholeText)
				continue
			assert xtext.nodeType == 1 # special or stylized element
			if xtext.tagName == "br":
				of.write("\\n")
				continue
			assert xtext.tagName == "span"
			color = COLORS[xtext.getAttribute("tts:color")]
			of.write("{\\c&H%s&}%s{\\r}" % (color, xtext.firstChild.wholeText))
		of.write("\n")
